compiler/bindings/python/iree/compiler/tools/onnx.py

```diff
@@ -0,0 +1,250 @@
+                        opset_version, options.min_opset_version)
+            try:
+                converted_model = onnx.version_converter.convert_version(
+                    model, options.min_opset_version)
+            except:
+                # Conversion failed. Do our best with the original file.
+                logger.warning("Converting onnx model opset version from %s "
+                               "to %s failed. Continuning without conversion.",
+                               opset_version, options.min_opset_version)
+                converted_model = model
+        else:
+            # No conversion needed.
+            converted_model = model
+
+        if options.entry_point_name:
+            converted_model.graph.name = options.entry_point_name
+
+        # import onnx model
+        model_proto = load_onnx_model(
+            converted_model, options.data_prop, options.data_dir)
+
+        logger.info("Importing graph: '%s' from onnx model",
+                    model_proto.graph.name)
+        context = Context()
+        model_info = onnx_importer.ModelInfo(model_proto)
+        module = model_info.create_module(context=context).operation
+
+        if options.module_name:
+            module.attributes["sym_name"] = StringAttr.get(
+                options.module_name, context)
+
+        importer = onnx_importer.NodeImporter.define_function(
+            model_info.main_graph, module)
+        importer.import_all()
+
+        if options.verify_module:
+            module.verify()
+
+        if options.use_bytecode:
+            with open(onnx_iree_input, "wb") as f:
+                module.write_bytecode(f)
+        else:
+            with open(onnx_iree_input, "wt", encoding="utf-8") as f:
+                print(module.get_asm(assume_verified=options.verify_module), file=f)
+
+        if options.import_only:
+            if options.output_file:
+                return None
+            if options.use_bytecode:
+                with open(onnx_iree_input, "rb") as f:
+                    return f.read()
+            else:
+                with open(onnx_iree_input, "rt", encoding="utf-8") as f:
+                    return f.read()
+
+        # compile onnx model
+        compile_cl = build_compile_command_line(onnx_iree_input, tfs, options)
+        result = invoke_pipeline([compile_cl])
+        if options.output_file:
+            return None
+        return result
+
+
+def compile_saved_model(model_path: IO[bytes] | str | os.PathLike, **kwargs) -> None | str | IO[bytes]:
+    """Import and compile an ONNX model.
+
+    Args:
+        model IO[bytes] | str | os.PathLike: The path for the ONNX model to
+            import and compile.
+
+    Returns:
+        None | str | IO[bytes]: If no output file is specified, the compiled
+            model as a string or bytes depending on `output_format` and `use_bytecode`.
+    """
+    model = onnx.load_model(model_path)
```

Comment:
I think this will fail if external data is not stored in the same directory as the model. 

After looking through this a bit more carefully, I have a few ideas. Sorry for the delay, and thanks for continuing to iterate on this!

0. Without exception, `load_onnx_model` should take in a filepath (+ options) and return an `onnx.ModelProto`, otherwise the naming is confusing. 
1. The pre-processing that occurs in `load_onnx_model`, in your current implementation, now requires more temp files if a user is starting from a filepath. Specifically, if the model is too large for in-memory shape inference, you additionally need to store the original onnx model to the temp directory even though it already exists on your device. These models can be absolutely huge, and this is the exact case which is being affected.
2. Following up on the previous two points, I think it makes the most sense to rename `load_onnx_model` to `preprocess_model` and have this function take either an `onnx.ModelProto`, or a filepath. If a filepath is provided, then it will load the model first, then apply the preprocessing steps. This way, we won't need to re-save the original onnx model in the specific case discussed in (1.).
3. All onnx model preprocessing that is required for importing to MLIR should happen in one function. Currently, this would likely be in `load_onnx_model`/`preprocess_model`. In particular, I think the opset version updating should likely occur there instead of in `compile_model`. This way, if someone just wants to update all of their onnx models to be "import-ready" (e.g. for a testing-suite) they can use that function on it's own.
4. Related to 3.: see if you can do opset version updating without loading external weights. This would be a great if possible.
5. With the above changes, I'd personally like `compile_model` to optionally take in a filepath or `ModelProto` as an input, since the preprocessing function takes either as well. 
6. Additionally, with `compile_model`, it might be good to add a bool option like `preprocess` with a default of `True`. If someone already preprocessed their models and doesn't want to redo shape inference etc., it would be nice if they could pass `preprocess=False` and jump straight to compilation.
7. Perhaps instead of 6., you could add a function `compile_preprocessed_model`, which assumes preprocessing has already occurred. Then call this inside `compile_model` after preprocessing. If someone wants to skip preprocessing they could just use the other function.

Let me know what you think about these ideas. I'm going to be out of office until 9/10/2024, but I'll try to check in on my phone when I can.